source/request_source/request_source_plugin_impl.h

```diff
@@ -0,0 +1,124 @@
+// config_factory =
+//     Envoy::Config::Utility::getAndCheckFactoryByName<RequestSourcePluginConfigFactory>(
+//         "nighthawk.stub-request-source-plugin");
+// RequestSourcePtr plugin =
+//     config_factory.createRequestSourcePlugin(config, std::move(api), std::move(header));
+
+class DummyRequestSourcePluginConfigFactory : public virtual RequestSourcePluginConfigFactory {
+public:
+  // This is a hardcoded string.
+  std::string name() const override;
+  // This returns an empty version of the expected StubPluginConfig from request_source_plugin.proto
+  Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+  // This is the primary method that is used to get RequestSources.
+  // This implementation is thread safe, but the RequestSource it generates doesn't do much.
+  RequestSourcePtr createRequestSourcePlugin(const Envoy::Protobuf::Message& message,
+                                             Envoy::Api::ApiPtr api,
+                                             Envoy::Http::RequestHeaderMapPtr header) override;
+};
+
+// This factory will be activated through RequestSourceFactory in factories.h
+DECLARE_FACTORY(DummyRequestSourcePluginConfigFactory);
+
+// Sample Request Source for small RequestOptionsLists. Loads a copy of the RequestOptionsList in
+// memory and replays them.
+// @param request_max The number of requests the requestGenerator produced by get() will generate. 0
+// means it is unlimited.
+// @param header the default header that will be overridden by values taken from the options_list,
+// any values not overridden will be used.
+// @param options_list A copy of the options_list will be loaded in memory. The RequestGenerator
+// produced by get() will use options from the options_list to overwrite values in the header, and
+// create new requests. if request_max is greater than the length of options_list, it will loop.
+// This is not thread safe.
+class RequestOptionsListRequestSource : public RequestSource {
+public:
+  explicit RequestOptionsListRequestSource(
+      const uint32_t request_max, Envoy::Http::RequestHeaderMapPtr header,
+      std::unique_ptr<nighthawk::client::RequestOptionsList> options_list);
+  // This get function is not thread safe, because multiple threads calling get simultaneously will
+  // result in a collision as it attempts to update its request_count_.
+  RequestGenerator get() override;
+
+  // default implementation
+  void initOnThread() override;
+
+private:
+  Envoy::Http::RequestHeaderMapPtr header_;
+  const std::unique_ptr<nighthawk::client::RequestOptionsList> options_list_;
+  std::vector<uint32_t> request_count_;
+  const uint32_t request_max_;
+};
+
+// Factory that creates a RequestOptionsListRequestSource from a FileBasedPluginConfig proto.
+// Registered as an Envoy plugin.
+// Implementation of RequestSourceConfigFactory which produces a RequestSource that keeps an
```

Comment:
This comment contains a mix of good class comments and implementation comments. We shouldn't comment or describe implementation here as it is bound to change and isn't part of the API contract. An example of implementation comments here are:

- the fact that the class keeps a RequestOptionsList in memory
- the fact that file_lock_ is used to protect the list.

On the other hand saying that the class is not thread safe is a very good comment here, but it doesn't need more details.

Please use your best judgment to adjust the comments so they talk about *what* a class does and how it can *be used*, but never about *how* it does its job.

Reply:
I'm not sure I'm doing a very good job at this, but I'll do my best.